cdap-api/src/main/java/co/cask/cdap/api/common/RuntimeArguments.java

```diff
@@ -74,37 +76,85 @@ private RuntimeArguments() {
   }
 
   /**
-   * Extract all arguments for a given scope.
+   * Extracts the scope from the argument as identified by its prefix.
+   *
+   * 1. Argument can be prefixed by "<scopetype>.<name>.". e.g. mapreduce.myMapReduce.read.timeout=30. In this case
+   * only MapReduce program named 'myMapReduce' will receive the argument read.timeout=30.
+   * 2. Argument can be prefixed by "<scopetype>.*.". e.g. mapreduce.*.read.timeout=30. In this case all the underlying
+   * MapReduce programs will receive the argument read.timeout=30.
+   * 3. If the argument is not prefixed with any scope, it is considered global and is passed further without any
+   * changes. e.g. read.timeout=30
+   *
+   * Arguments scoped by DATASET are only filtered while inside the dataset scope. They are passed from other
```

Comment:
We have to be careful about filtering in general. For example, the dataset framework uses a lot of properties prefixed with "dataset.". That means we would always filter these out. I would not be suprised if Hadoop has built-in property names prefixed with "mapreduce.", or Spark has built-in property names prefixed with "spark.". All these would be filtered out? Because they look as if they have a scope but they don't really have one in the sense of this class. 

Perhaps we should not filter anything. But instead simply find all the properties that have a matching prefix, and add them to the arguments without the prefix. But keep the existing ones, including matching ones and non-matching ones. It means there will be a bunch of properties in the runtime arguments that are irrelevant and that do not really have to be there. But it is better to include some irrelevant ones than to exclude some relevant ones.